File: src/main/java/com/soze/r2d/R2D.java
package com.soze.r2d;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.scenes.scene2d.Actor;
import com.badlogic.gdx.scenes.scene2d.Event;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Cell;
import com.badlogic.gdx.scenes.scene2d.ui.Dialog;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextField;
import com.badlogic.gdx.scenes.scene2d.ui.Window;
import com.badlogic.gdx.utils.Align;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;
import java.util.function.BiConsumer;
import java.util.function.Function;

public class R2D {

  private static final Logger LOG = LoggerFactory.getLogger(R2D.class);
  private static int RENDERS = 0;

  private static Runnable stateChangeCallback = () -> {
  };
  private static R2DNode currentVDom = new R2DNode();


  public static void render(Element element, Group group) {
    long time0 = System.nanoTime();
    LOG.debug("First render");
    RENDERS++;

    Actor focus = null;
    if (group.getStage() != null) {
      focus = group.getStage().getKeyboardFocus();
    }
    stateChangeCallback = () -> {
      LOG.debug("stateChangeCallback called, re-rendering everything!");
      render(element, group);
    };
    renderChild(currentVDom, element, group);
    if (group.getStage() != null) {
      group.getStage().setKeyboardFocus(focus);
    }
    long totalTime = System.nanoTime() - time0;
    LOG.debug("{} ms to render", TimeUnit.NANOSECONDS.toMillis(totalTime));
    LOG.debug("Rendered {} times", RENDERS);
  }

  private static void renderChild(R2DNode vdom, Element element, Group group) {
    vdom.setGroup(group);
    Object type = element.getType();
    boolean differentType = vdom.getType() != type;
    if (differentType) {
      group.clearChildren();
      vdom.getChildren().clear();
    }
    vdom.setType(type);
    if (type instanceof String) {
      renderStringType(vdom, element, group, differentType);
    }
    if (type instanceof Class<?>) {
      renderComponentType(vdom, element, group, differentType);
    }
  }

  private static void renderStringType(R2DNode vdom, Element element, Group group, boolean differentType) {
    UiState props = element.getProps();
    boolean show = props.get("show", true);
    Actor actor = getActor(vdom, element, differentType);
    if (!show) {
      group.removeActor(actor);
      return;
    }
    
    vdom.setActor(actor);
    
    if (group instanceof Table) {
      Table table = (Table) group;
      boolean row = props.get("row", false);
      if (row) {
        table.row();
      }
      Cell cell = table.add(actor);
      int colspan = (int) props.get("colspan", 1);
      cell.colspan(colspan);

    } else {
      group.addActor(actor);
    }
    if (actor instanceof Group) {
      Group currentGroup = (Group) actor;
      currentGroup.clearChildren();
      List<Element> children = (List<Element>) props.get("children");
      List<R2DNode> nextChildren = new ArrayList(children.size());
      for (Element child : children) {
        R2DNode node = new R2DNode();
        node.setType("VOID");
        nextChildren.add(node);
      }
      List<R2DNode> previousChildren = vdom.getChildren();
      for (int i = 0; i < children.size(); i++) {
        Element childElement = children.get(i);
        R2DNode currentChild = previousChildren.size() > i ? previousChildren.get(i) : null;
        R2DNode nextVDom;

        //if current child is null, we are creating a new one
        if (currentChild == null) {
          nextVDom = new R2DNode();
          nextVDom.setType(childElement.getType());
          nextVDom.setParent(vdom);
        } else {
          // a child was present already, we need to check if it's the same or different
          boolean differentChildType = childElement.getType() != currentChild.getType();
          if (!differentChildType) {
            nextVDom = currentChild;
          } else {
            nextVDom = new R2DNode();
            nextVDom.setType(childElement.getType());
            nextVDom.setParent(vdom);
          }
        }
        nextChildren.set(i, nextVDom);
        renderChild(nextVDom, childElement, currentGroup);
      }

      vdom.getChildren().clear();
      vdom.getChildren().addAll(nextChildren);
    }

  }

  private static Actor getActor(R2DNode vdom, Element element, boolean differentType) {
    if (!differentType && vdom.getActor() != null) {
      return applyProps(element.getProps(), vdom.getActor());
    }
    return createActor((String) element.getType(), element.getProps());
  }

  private static void renderComponentType(R2DNode vdom, Element element, Group group, boolean differentType) {
    Component component = getComponent(vdom, element, differentType);
    component.setStateChangeCallback(R2D.stateChangeCallback);
    Element nextElement = component.render();
    vdom.setComponent(component);
    R2DNode nextVdom = getNextVdomComponent(vdom, nextElement, differentType);
    vdom.getChildren().clear();
    vdom.getChildren().add(nextVdom);
    renderChild(nextVdom, nextElement, group);
  }

  private static Component getComponent(R2DNode vdom, Element element, boolean differentType) {
    if (!differentType && vdom.getComponent() != null) {
      return vdom.getComponent();
    }
    return createComponent((Class<Component>) element.getType(), element.getProps());
  }

  private static R2DNode getNextVdomComponent(R2DNode vdom, Element nextElement, boolean differentType) {
    if (!differentType && !vdom.getChildren().isEmpty()) {
      return vdom.getChildren().get(0);
    }
    R2DNode nextVDom = new R2DNode();
    nextVDom.setType(nextElement.getType());
    nextVDom.setParent(vdom);
    return nextVDom;
  }

  private static Actor createActor(String type, UiState props) {
    Actor actor = null;
    switch (type) {
      case "LABEL":
        actor = createLabel(props);
        break;
      case "TABLE":
        actor = createTable(props);
        break;
      case "TEXT_FIELD":
        actor = createTextField(props);
        break;
      case "BUTTON":
        actor = createButton(props);
        break;
      case "WINDOW":
        actor = createWindow(props);
        break;
      case "DIALOG":
        actor = createDialog(props);
        break;
      default:
        actor = new Table();
    }
    return applyPropsActor(props, actor);
  }

  private static Actor createLabel(UiState props) {
    Label.LabelStyle style = new Label.LabelStyle(new BitmapFont(), Color.WHITE);
    return applyProps(props, new Label(null, style));
  }

  private static Actor createTable(UiState props) {
    return applyProps(props, new Table());
  }

  private static Actor createTextField(UiState props) {
    TextField.TextFieldStyle style = new TextField.TextFieldStyle();
    style.font = new BitmapFont();
    return applyProps(props, new TextField(null, style));
  }

  private static Actor applyProps(UiState props, Actor actor) {
    Class<?> clazz = actor.getClass();
    applyPropsActor(props, actor);
    if (clazz.isAssignableFrom(Table.class)) {
      return applyProps(props, (Table) actor);
    }
    if (clazz.isAssignableFrom(TextField.class)) {
      return applyProps(props, (TextField) actor);
    }
    if (clazz.isAssignableFrom(Label.class)) {
      return applyProps(props, (Label) actor);
    }
    if (clazz.isAssignableFrom(Dialog.class)) {
      return applyProps(props, (Dialog) actor);
    }
    return actor;
  }

  private static Table applyProps(UiState props, Table table) {
    table.setFillParent(props.get("fillParent", false));
    return table;
  }

  private static TextField applyProps(UiState props, TextField textField) {
    TextField.TextFieldStyle defaultStyle = new TextField.TextFieldStyle();
    defaultStyle.font = new BitmapFont();
    defaultStyle.fontColor = Color.WHITE;
    defaultStyle.cursor = DefaultCursor.getCursor();
    defaultStyle.selection = DefaultCursor.getCursor();

    TextField.TextFieldStyle style = props.get("textFieldStyle", defaultStyle);
    String text = props.get("text", "");
    textField.setText(text);
    textField.setStyle(style);

    BiConsumer<TextField, Character> onChange = (BiConsumer<TextField, Character>) props.get("onChange");
    if (onChange != null) {
      textField.setTextFieldListener(onChange::accept);
    }
    return textField;
  }

  private static Label applyProps(UiState props, Label label) {
    Label.LabelStyle defaultStyle = new Label.LabelStyle();
    defaultStyle.font = new BitmapFont();
    defaultStyle.fontColor = Color.WHITE;
    String text = props.get("text", "");
    Label.LabelStyle style = props.get("labelStyle", defaultStyle);
    label.setText(text);
    label.setStyle(style);
    boolean disabled = props.get("disabled", false);
    if (disabled) {
      style.fontColor = Color.GRAY;
    }
    Function<Event, Boolean> onClick = (Function<Event, Boolean>) props.get("onClick");
    if (label.getListeners().size > 0) {
      label.getListeners().removeIndex(0);
    }
    if (onClick != null) {
      label.addListener((event) -> {
        if (!disabled && event instanceof InputEvent && ((InputEvent) event).getType() == InputEvent.Type.touchDown) {
          return onClick.apply(event);
        }
        return false;
      });
    }
    return label;
  }
  
  private static Dialog applyProps(UiState props, Dialog dialog) {
    dialog.setPosition(props.get("x", 0f), props.get("y", 0f), Align.center);
    dialog.getTitleLabel().setText(props.get("title", ""));
    
    return dialog;
  }

  private static Actor applyPropsActor(UiState props, Actor actor) {
    actor.setName((String) props.get("name"));
    return actor;
  }

  private static Actor createButton(UiState props) {
    LOG.warn("CREATING NON IMPLEMENTED BUTTON");
    return new Table();
  }
  
  private static Actor createWindow(UiState props) {
    LOG.warn("CREATING NON IMPLEMENTED WINDOW");
    return new Table();
  }
  
  private static Actor createDialog(UiState props) {
    Window.WindowStyle style = new Window.WindowStyle();
    style.titleFont = new BitmapFont();
    style.titleFontColor = Color.WHITE;
    
    Dialog dialog = new Dialog(props.get("title", ""), style);
    return applyProps(props, dialog);
  }

  private static Component createComponent(Class<Component> clazz, UiState props) {
    try {
      Constructor<Component> constructor = clazz.getDeclaredConstructor(props.getClass());
      return constructor.newInstance(props);
    } catch (NoSuchMethodException e) {
      e.printStackTrace();
    } catch (IllegalAccessException e) {
      e.printStackTrace();
    } catch (InstantiationException e) {
      e.printStackTrace();
    } catch (InvocationTargetException e) {
      e.printStackTrace();
    }
    throw new IllegalStateException(clazz + " does not have a proper constructor");
  }

}

File: src/main/kotlin/com/soze/r2d/R2D.kt
package com.soze.lifegame.uibuilder

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.scenes.scene2d.Actor
import com.badlogic.gdx.scenes.scene2d.Event
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.Cell
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextField
import com.badlogic.gdx.utils.DelayedRemovalArray
import com.soze.r2d.DefaultCursor
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.lang.IllegalStateException
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.reflect.KClass
import kotlin.reflect.full.primaryConstructor

fun com.badlogic.gdx.scenes.scene2d.ui.Table.lastCell(): Cell<*>? {
  val cells = this.cells
  return if (cells.size > 0) cells.get(cells.size - 1) else null
}

/**
 * Renders R elements into a Scene2D structure.
 */
object R2D {

  private var currentVDom: R2DNode = R2DNode("INITIAL")
  private var LOG: Logger = LoggerFactory.getLogger(R2D::class.java)

  private var stateChangeCallback: () -> Unit = fun() {}

  @JvmStatic
  fun render(element: Element, group: Group) {
    val time0 = System.nanoTime()
    LOG.debug("First render")
    val focus = group.stage?.keyboardFocus
    R2D.stateChangeCallback = fun() {
      LOG.debug("stateChangeCallback called, re-rendering everything!")
      render(element, group)
    }
    renderChild(currentVDom, element, group)

    group.stage?.keyboardFocus = focus
    val totalTime = System.nanoTime() - time0
    LOG.debug("${TimeUnit.NANOSECONDS.toMillis(totalTime)} ms to render")
  }

  private fun renderChild(vdom: R2DNode, element: Element, group: Group) {
    vdom.group = group
    val type = element.type
    var differentType = vdom.type != type
    if (differentType) {
      group.clearChildren()
      vdom.children.clear()
    }
    vdom.type = type
    if (type is String) {
      renderStringType(vdom, element, group, differentType)
    }
    if (type is KClass<*> || type is Class<*>) {
      renderComponentType(vdom, element, group, differentType)
    }
  }

  private fun renderStringType(vdom: R2DNode, element: Element, group: Group, differentType: Boolean) {
    val props = element.props
    val actor = getActor(vdom, element, differentType)
    vdom.actor = actor

    if (group is Table) {
      val row = props["row"] as Boolean? ?: false
      if (row) {
        group.lastCell()?.let {
          it.row()
        }
      }
      group.add(actor).apply {
        val colspan = props["colspan"] as Int? ?: 1
        colspan(colspan)
      }
    } else {
      group.addActor(actor)
    }
    if (actor is Group) {
      actor.clearChildren()
      val children = props["children"] as List<Element>
      val nextChildren: MutableList<R2DNode> = ArrayList(children.size)
      for (i in 0 until children.size) {
        nextChildren.add(R2DNode("VOID"))
      }
      val previousChildren = vdom.children
      children.forEachIndexed { index, element ->
        var currentChild: R2DNode? = if (previousChildren.size > index) previousChildren[index] else null
        var nextVdom: R2DNode?
        //if current child is null, we are creating a new one
        if (currentChild == null) {
          nextVdom = R2DNode(type = element.type, parent = vdom)
        } else {
          // a child was present already, we need to check if it's the same or different
          val differentChildType = element.type != currentChild.type
          //if type is the same
          if (!differentChildType) {
            nextVdom = currentChild
          } else {
            nextVdom = R2DNode(type = element.type, parent = vdom)
          }
        }
        nextChildren[index] = nextVdom
        renderChild(nextVdom, element, actor)
      }
      vdom.children = nextChildren
    }

  }

  private fun getActor(vdom: R2DNode, element: Element, differentType: Boolean): Actor {
    if (!differentType && vdom.actor != null) {
      return applyProps(element.props, vdom.actor!!)
    }
    return createActor(element.type as String, element.props)
  }

  private fun renderComponentType(vdom: R2DNode, element: Element, group: Group, differentType: Boolean) {
    val component = getComponent(vdom, element, differentType)
    component.stateChangeCallback = R2D.stateChangeCallback
    val nextElement = component.render()
    vdom.component = component
    val nextVdom = getNextVdomComponent(vdom, nextElement, differentType)
    vdom.children.clear()
    vdom.children.add(nextVdom)
    renderChild(nextVdom, nextElement, group)
  }

  private fun getComponent(vdom: R2DNode, element: Element, differentType: Boolean): Component {
    if (!differentType && vdom.component != null) {
      return vdom.component!!
    }
    return createComponent(element.type, element.props)
  }

  private fun getNextVdomComponent(vdom: R2DNode, nextElement: Element, differentType: Boolean): R2DNode {
    if (!differentType && vdom.children.isNotEmpty()) {
      return vdom.children[0]
    }
    return R2DNode(type = nextElement.type, parent = vdom)
  }

  private fun createActor(type: String, props: UiState): Actor {
    val actor = when (type) {
      "LABEL" -> createLabel(props)
      "TABLE" -> createTable(props)
      "TEXT_FIELD" -> createTextField(props)
      "BUTTON" -> createButton(props)
      else -> Table()
    }
    return applyPropsActor(props, actor)
  }

  private fun createLabel(props: UiState): Actor {
    val style = Label.LabelStyle(BitmapFont(), Color.WHITE)
    return applyProps(props, Label(null, style))
  }

  private fun createTable(props: UiState): Actor {
    return applyProps(props, Table())
  }

  private fun createTextField(props: UiState): Actor {
    val style = TextField.TextFieldStyle()
    style.font = BitmapFont()
    return applyProps(props, TextField(null, style))
  }

  private fun applyProps(props: UiState, actor: Actor): Actor {
    val clazz = actor::class.java
    applyPropsActor(props, actor)
    return when (clazz) {
      Table::class.java -> applyProps(props, actor as Table)
      TextField::class.java -> applyProps(props, actor as TextField)
      Label::class.java -> applyProps(props, actor as Label)
      else -> actor
    }
  }

  private fun applyProps(props: UiState, table: Table): Table {
    table.setFillParent(props["fillParent"] as Boolean? ?: false)
    return table
  }

  private fun applyProps(props: UiState, textField: TextField): TextField {
    val defaultStyle = TextField.TextFieldStyle()
    defaultStyle.font = BitmapFont()
    defaultStyle.fontColor = Color.WHITE
    defaultStyle.cursor = DefaultCursor.get()
    defaultStyle.selection = DefaultCursor.get()

    val style = props["textFieldStyle"] as TextField.TextFieldStyle? ?: defaultStyle
    val text = props["text"] as String? ?: ""
    textField.text = text
    textField.style = style

    val onChange = props["onChange"] as ((field: TextField, char: Char) -> Unit)?
    onChange?.let {
      textField.setTextFieldListener(onChange)
    }
    return textField
  }

  private fun applyProps(props: UiState, label: Label): Label {
    val defaultStyle = Label.LabelStyle()
    defaultStyle.font = BitmapFont()
    defaultStyle.fontColor = Color.WHITE
    val text = props["text"] as String? ?: ""
    val style = props["labelStyle"] as Label.LabelStyle? ?: defaultStyle
    label.setText(text)
    label.style = style
    val disabled = props["disabled"] as Boolean? ?: false
    if (disabled) {
      style.fontColor = Color.GRAY
    }
    val onClick = props["onClick"] as ((event: Event) -> Boolean)?
    if (label.listeners.size > 0) {
      (label.listeners as DelayedRemovalArray).removeIndex(0)
    }
    onClick?.let {
      label.addListener(fun(event: Event): Boolean {
        if (!disabled && event is InputEvent && event.type == InputEvent.Type.touchDown) {
          return onClick(event)
        }
        return false
      })
    }
    return label
  }

  private fun applyPropsActor(props: UiState, actor: Actor): Actor {
    actor.name = props["name"] as String?
    return actor
  }

  private fun createButton(props: UiState): Actor {
    LOG.warn("CREATING NON IMPLEMENTED BUTTON")
    return Table()
  }

  private fun createComponent(clazz: Any, props: UiState): Component {
    if (clazz is Class<*>) {
      val declaredConstructor = clazz.getDeclaredConstructor(UiState::class.java)
      return declaredConstructor.newInstance(props) as Component
    }
    if (clazz is KClass<*>) {
      val primaryConstructor = clazz.primaryConstructor!!
      return primaryConstructor.call(props) as Component
    }
    throw IllegalStateException("$clazz is not a class")
  }


}